runtime/common/ArgumentConversion.cpp

```diff
@@ -100,6 +92,75 @@ static Value genConstant(OpBuilder &, cudaq::cc::StructType, void *,
 static Value genConstant(OpBuilder &, cudaq::cc::ArrayType, void *,
                          ModuleOp substMod, llvm::DataLayout &);
 
+static Value genConstant(OpBuilder &builder, const cudaq::state *v,
+                         ModuleOp substMod, llvm::DataLayout &layout,
+                         llvm::StringRef kernelName, bool isSimulator) {
+  if (isSimulator) {
+    // The program is executed remotely, materialize the simulation data
+    // into an array and create a new state from it.
+    auto numQubits = v->get_num_qubits();
+
+    // We currently only synthesize small states.
+    if (numQubits > 14) {
+      TODO("large (>14 qubit) cudaq::state* argument synthesis for simulators");
+      return {};
+    }
+
+    auto size = 1ULL << numQubits;
+    auto ctx = builder.getContext();
+    auto loc = builder.getUnknownLoc();
+    auto is64Bit =
+        v->get_precision() == cudaq::SimulationState::precision::fp64;
+    auto eleTy = is64Bit ? ComplexType::get(Float64Type::get(ctx))
+                         : ComplexType::get(Float32Type::get(ctx));
+    auto arrTy = cudaq::cc::ArrayType::get(ctx, eleTy, size);
+    static unsigned counter = 0;
+    auto ptrTy = cudaq::cc::PointerType::get(arrTy);
+
+    cudaq::IRBuilder irBuilder(ctx);
+    auto genConArray = [&]<typename T>() -> Value {
+      auto data = std::make_unique<std::complex<T>[]>(size);
+      for (std::size_t i = 0; i < size; i++) {
+        data[i] = (*v)({i}, 0);
+      }
+      std::string name =
+          kernelName.str() + ".rodata_synth_" + std::to_string(counter++);
+      std::vector<std::complex<T>> vec(data.get(), data.get() + size);
```

Comment:
I could create a vector and push_back things to it - maybe it will look better, but we will still allocate memory... Will do that.

Reply:
In the current code, we are allocating 2x the memory (once for `data` and once for `vec`). I think it could be reduced to a single allocation, and push_back could be avoided, if done like this:
```cpp
std::vector<std::complex<T>> vec(size);
for (std::size_t i = 0; i < size; i++)
  vec[i] = (*v)({i}, 0);
```

Reply:
In other words, `std::vector<std::complex<T>> vec(data.get(), data.get() + size);` is simply making a copy of `data`, not just pointing to its underlying data, so that's where the duplicate allocation is coming from.

Reply:
Thanks! Updated.